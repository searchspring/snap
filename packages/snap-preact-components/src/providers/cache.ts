export { CacheProvider } from '@emotion/react';
import createCache from '@emotion/cache';

export const cache = createCache({
	key: 'ss',
	prepend: true,
});
