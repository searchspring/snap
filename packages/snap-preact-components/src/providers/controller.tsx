import { jsx } from '@emotion/react';
import preact, { h, createContext, ComponentChildren, ComponentType } from 'preact';
import { useContext } from 'preact/hooks';
import type { AbstractController } from '@searchspring/snap-controller';

const ControllerContext = createContext<null | AbstractController>(null);

export const ControllerProvider = ({ children, controller }: { children: ComponentChildren; controller: AbstractController }) => {
	return <ControllerContext.Provider value={controller}>{children}</ControllerContext.Provider>;
};

export const useController = () => useContext(ControllerContext);

export const withController = (Component: ComponentType) => (props: any) => {
	return <Component controller={useController()} {...props} />;
};
