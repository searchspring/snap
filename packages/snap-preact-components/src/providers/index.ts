export * from './store';
export * from './theme';
