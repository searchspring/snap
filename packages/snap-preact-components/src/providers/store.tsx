import preact, { h, createContext, ComponentChildren, ComponentType } from 'preact';
import { useContext } from 'preact/hooks';
import type { AbstractStore } from '@searchspring/snap-store-mobx';

const StoreContext = createContext<null | AbstractStore>(null);

export const StoreProvider = ({ children, store }: { children: ComponentChildren; store: AbstractStore }) => {
	return <StoreContext.Provider value={store}>{children}</StoreContext.Provider>;
};

export const useStore = () => useContext(StoreContext);

export function withStore<C extends ComponentType>(Component: C): C {
	return ((props: any) => <StoreContext.Consumer>{(store) => <Component {...props} store={store} />}</StoreContext.Consumer>) as C;
}
