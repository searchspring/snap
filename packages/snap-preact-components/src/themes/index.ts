export * from './pike';
export * from './bocachica';
