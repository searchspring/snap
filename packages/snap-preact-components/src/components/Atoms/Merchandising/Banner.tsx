/** @jsx jsx */
import { h } from 'preact';

import { jsx, css } from '@emotion/react';
import classnames from 'classnames';

import { BannerContent, BannerType, ComponentProps } from '../../../types';
import { Theme, useTheme } from '../../../providers';

const CSS = {
	banner: ({ style }) =>
		css({
			'& iframe': {
				maxWidth: '100%',
			},
			...style,
		}),
};

export function Banner(properties: BannerProps): JSX.Element {
	const globalTheme: Theme = useTheme();

	const props: BannerProps = {
		// default props
		content: [],
		type: '',
		// global theme
		...globalTheme?.components?.banner,
		// props
		...properties,
		...properties.theme?.components?.banner,
	};

	const { content, type, disableStyles, className, style } = props;

	if (type === BannerType.INLINE) {
		console.warn(`BannerType '${BannerType.INLINE}' is not supported in <Banner /> component`);
		return;
	}

	return (
		content &&
		content[type]?.length && (
			<div
				className={classnames('ss__banner', `ss__banner--${type}`, className)}
				css={!disableStyles && CSS.banner({ style })}
				dangerouslySetInnerHTML={{
					__html: content[props.type].join(''),
				}}
			/>
		)
	);
}

export interface BannerProps extends ComponentProps {
	content: BannerContent | [];
	type: BannerType;
}
