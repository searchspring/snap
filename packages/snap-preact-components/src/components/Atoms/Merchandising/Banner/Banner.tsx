/** @jsxRuntime classic */
/** @jsx jsx */
import { Fragment, h } from 'preact';

import { jsx, css } from '@emotion/react';
import classnames from 'classnames';

import { ComponentProps, StylingCSS } from '../../../../types';
import { Theme, useTheme, CacheProvider } from '../../../../providers';
import { BannerContent, ContentType } from '@searchspring/snap-store-mobx';

const CSS = {
	banner: () =>
		css({
			'& iframe, img': {
				maxWidth: '100%',
				height: 'auto',
			},
		}),
};

export function Banner(properties: BannerProps): JSX.Element {
	const globalTheme: Theme = useTheme();

	const props: BannerProps = {
		// global theme
		...globalTheme?.components?.banner,
		// props
		...properties,
		...properties.theme?.components?.banner,
	};

	const { content, type, disableStyles, className, style } = props;

	if (type === ContentType.INLINE) {
		console.warn(`BannerType '${ContentType.INLINE}' is not supported in <Banner /> component`);
		return <Fragment></Fragment>;
	}
	const styling: { css?: StylingCSS } = {};
	if (!disableStyles) {
		styling.css = [CSS.banner(), style];
	} else if (style) {
		styling.css = [style];
	}

	let bannerContent;
	if (content && content[type]) {
		bannerContent = content[type] as string[];
	}

	return bannerContent && bannerContent.length ? (
		<CacheProvider>
			<div
				className={classnames('ss__banner', `ss__banner--${type}`, className)}
				{...styling}
				dangerouslySetInnerHTML={{
					__html: bannerContent.join(''),
				}}
			/>
		</CacheProvider>
	) : (
		<Fragment></Fragment>
	);
}

export interface BannerProps extends ComponentProps {
	content: BannerContent;
	type: ContentType;
}
