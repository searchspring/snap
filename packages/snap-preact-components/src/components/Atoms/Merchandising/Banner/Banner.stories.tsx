import { h } from 'preact';

import { ArgsTable, PRIMARY_STORY, Markdown } from '@storybook/blocks';

import { Banner, BannerProps } from './Banner';
import { componentArgs, highlightedCode } from '../../../../utilities';
import { Snapify } from '../../../../utilities/snapify';
import Readme from './readme.md';
import { SearchController } from '@searchspring/snap-controller';

export default {
	title: 'Atoms/Banner',
	component: Banner,
	tags: ['autodocs'],
	parameters: {
		docs: {
			page: () => (
				<div>
					<Markdown
						options={{
							overrides: {
								code: highlightedCode,
							},
						}}
					>
						{Readme}
					</Markdown>
					<ArgsTable story={PRIMARY_STORY} />
				</div>
			),
		},
	},
	argTypes: {
		content: {
			description: 'Banner content store reference',
			type: { required: true },
			table: {
				type: {
					summary: 'banner content store object',
				},
			},
			control: { type: 'none' },
		},
		type: {
			description: 'Banner position type',
			type: { required: true },
			table: {
				type: {
					summary: 'string',
				},
			},
			options: ['header', 'footer', 'left', 'inline', 'banner'],
			control: {
				type: 'select',
			},
		},
		...componentArgs,
	},
};

const snapInstance = Snapify.search({ id: 'Banner', globals: { siteId: '8uyt2m', search: { query: { string: 'glasses' } } } });

export const Header = (args: BannerProps, { loaded: { controller } }: { loaded: { controller: SearchController } }) => (
	<Banner {...args} content={controller?.store?.merchandising?.content} />
);
Header.loaders = [
	async () => {
		await snapInstance.search();
		return {
			controller: snapInstance,
		};
	},
];
Header.args = {
	type: 'header',
};

export const Footer = (args: BannerProps, { loaded: { controller } }: { loaded: { controller: SearchController } }) => (
	<Banner {...args} content={controller?.store?.merchandising?.content} />
);
Footer.loaders = [
	async () => {
		await snapInstance.search();
		return {
			controller: snapInstance,
		};
	},
];
Footer.args = {
	type: 'footer',
};

export const Secondary = (args: BannerProps, { loaded: { controller } }: { loaded: { controller: SearchController } }) => (
	<Banner {...args} content={controller?.store?.merchandising?.content} />
);
Secondary.loaders = [
	async () => {
		await snapInstance.search();
		return {
			controller: snapInstance,
		};
	},
];
Secondary.args = {
	type: 'banner',
};

export const Left = (args: BannerProps, { loaded: { controller } }: { loaded: { controller: SearchController } }) => (
	<Banner {...args} content={controller?.store?.merchandising?.content} />
);
Left.loaders = [
	async () => {
		await snapInstance.search();
		return {
			controller: snapInstance,
		};
	},
];
Left.args = {
	type: 'left',
};
