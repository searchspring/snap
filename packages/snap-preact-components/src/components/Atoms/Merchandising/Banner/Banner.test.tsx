import { h } from 'preact';

import { render } from '@testing-library/preact';

import { Banner } from './Banner';
import { ContentType } from '@searchspring/snap-store-mobx';
import { ThemeProvider } from '../../../../providers';
import themes from '../../../../themes';

import { MockData } from '@searchspring/snap-shared';
import { SearchResponseModel } from '@searchspring/snapi-types';

const mockData = new MockData();
mockData.updateConfig({ search: 'merchandising' });
const searchResponse: SearchResponseModel = mockData.search();

describe('Merchandising Banner Component', () => {
	Object.keys(themes || {}).forEach((themeName) => {
		it(`uses ${themeName} theme`, () => {
			const theme = themes[themeName as keyof typeof themes];
			const rendered = render(<Banner theme={theme} content={searchResponse.merchandising?.content!} type={ContentType.BANNER} />);
			expect(rendered.asFragment()).toMatchSnapshot();
		});
	});

	const theme = {
		components: {
			banner: {
				style: {
					backgroundColor: 'red',
				},
			},
		},
	};
	const typesWithContent: ContentType[] = [ContentType.HEADER, ContentType.FOOTER, ContentType.BANNER];

	typesWithContent.forEach((type) => {
		it(`renders type:${type} banner`, () => {
			const rendered = render(<Banner content={searchResponse.merchandising?.content!} type={type} />);
			const merchBannerElement = rendered.container.querySelector(`.ss__banner.ss__banner--${type}`);
			expect(merchBannerElement).toBeInTheDocument();
			expect(merchBannerElement?.innerHTML).toBe(searchResponse.merchandising?.content![type]!.join(''));
			expect(rendered.asFragment()).toMatchSnapshot();
		});
	});

	it('doesnt render if empty', () => {
		// @ts-ignore - testing empty banner
		const rendered = render(<Banner content={[]} type={ContentType.LEFT} />);
		const merchBannerElement = rendered.container.querySelector('.ss__banner.ss__banner--left');
		expect(merchBannerElement).not.toBeInTheDocument();
		expect(rendered.asFragment()).toMatchSnapshot();
	});

	it('can disable styling', () => {
		const rendered = render(<Banner disableStyles={true} content={searchResponse.merchandising?.content!} type={ContentType.BANNER} />);
		const loadingbarElement = rendered.container.querySelector('.ss__banner');
		expect(loadingbarElement?.classList.length).toBe(2);
		expect(rendered.asFragment()).toMatchSnapshot();
	});

	it('renders with classname', () => {
		const className = 'classy';
		const rendered = render(<Banner className={className} content={searchResponse.merchandising?.content!} type={ContentType.BANNER} />);
		const merchBannerElement = rendered.container.querySelector('.ss__banner.ss__banner--banner');
		expect(merchBannerElement).toBeInTheDocument();
		expect(merchBannerElement).toHaveClass(className);
		expect(rendered.asFragment()).toMatchSnapshot();
	});

	it('is themeable with ThemeProvider', () => {
		const args = {
			content: searchResponse.merchandising?.content!,
			type: ContentType.BANNER,
		};
		const rendered = render(
			<ThemeProvider theme={theme}>
				<Banner {...args} />
			</ThemeProvider>
		);
		const bannerElement = rendered.container.querySelector('.ss__banner')!;
		const styles = getComputedStyle(bannerElement);
		expect(styles.backgroundColor).toBe(theme.components.banner.style.backgroundColor);
		expect(rendered.asFragment()).toMatchSnapshot();
	});

	it('is themeable with theme prop', () => {
		const args = {
			content: searchResponse.merchandising?.content!,
			type: ContentType.BANNER,
		};
		const rendered = render(<Banner {...args} theme={theme} />);
		const bannerElement = rendered.container.querySelector('.ss__banner')!;
		const styles = getComputedStyle(bannerElement);
		expect(styles.backgroundColor).toBe(theme.components.banner.style.backgroundColor);
		expect(rendered.asFragment()).toMatchSnapshot();
	});

	it('is themeable with theme prop overrides ThemeProvider', () => {
		const args = {
			content: searchResponse.merchandising?.content!,
			type: ContentType.BANNER,
		};
		const themeOverride = {
			components: {
				banner: {
					style: {
						backgroundColor: 'blue',
					},
				},
			},
		};
		const rendered = render(
			<ThemeProvider theme={theme}>
				<Banner {...args} theme={themeOverride} />
			</ThemeProvider>
		);
		const bannerElement = rendered.container.querySelector('.ss__banner')!;
		const styles = getComputedStyle(bannerElement);
		expect(styles.backgroundColor).toBe(themeOverride.components.banner.style.backgroundColor);
		expect(rendered.asFragment()).toMatchSnapshot();
	});
});
