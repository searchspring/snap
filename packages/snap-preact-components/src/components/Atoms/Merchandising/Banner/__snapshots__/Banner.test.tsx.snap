// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Merchandising Banner Component can disable styling 1`] = `
<DocumentFragment>
  <div
    class="ss__banner ss__banner--banner"
  >
    <script
      data-banner-html="Secondary banner."
      data-banner-type="html"
      type="text/widget"
    />
    Secondary banner.
    <script
      data-banner-html="Secondary banner."
      data-banner-type="html"
      type="text/widget"
    />
    Secondary banner again.
    <script
      data-banner-html="Secondary banner."
      data-banner-type="html"
      type="text/widget"
    />
    Secondary banner a third.
  </div>
</DocumentFragment>
`;

exports[`Merchandising Banner Component doesnt render if empty 1`] = `<DocumentFragment />`;

exports[`Merchandising Banner Component is themeable with ThemeProvider 1`] = `
<DocumentFragment>
  .emotion-0 {
  background-color: red;
}

.emotion-0 iframe,
.emotion-0 img {
  max-width: 100%;
  height: auto;
}

<div
    class="ss__banner ss__banner--banner emotion-0"
  >
    <script
      data-banner-html="Secondary banner."
      data-banner-type="html"
      type="text/widget"
    />
    Secondary banner.
    <script
      data-banner-html="Secondary banner."
      data-banner-type="html"
      type="text/widget"
    />
    Secondary banner again.
    <script
      data-banner-html="Secondary banner."
      data-banner-type="html"
      type="text/widget"
    />
    Secondary banner a third.
  </div>
</DocumentFragment>
`;

exports[`Merchandising Banner Component is themeable with theme prop 1`] = `
<DocumentFragment>
  .emotion-0 {
  background-color: red;
}

.emotion-0 iframe,
.emotion-0 img {
  max-width: 100%;
  height: auto;
}

<div
    class="ss__banner ss__banner--banner emotion-0"
  >
    <script
      data-banner-html="Secondary banner."
      data-banner-type="html"
      type="text/widget"
    />
    Secondary banner.
    <script
      data-banner-html="Secondary banner."
      data-banner-type="html"
      type="text/widget"
    />
    Secondary banner again.
    <script
      data-banner-html="Secondary banner."
      data-banner-type="html"
      type="text/widget"
    />
    Secondary banner a third.
  </div>
</DocumentFragment>
`;

exports[`Merchandising Banner Component is themeable with theme prop overrides ThemeProvider 1`] = `
<DocumentFragment>
  .emotion-0 {
  background-color: blue;
}

.emotion-0 iframe,
.emotion-0 img {
  max-width: 100%;
  height: auto;
}

<div
    class="ss__banner ss__banner--banner emotion-0"
  >
    <script
      data-banner-html="Secondary banner."
      data-banner-type="html"
      type="text/widget"
    />
    Secondary banner.
    <script
      data-banner-html="Secondary banner."
      data-banner-type="html"
      type="text/widget"
    />
    Secondary banner again.
    <script
      data-banner-html="Secondary banner."
      data-banner-type="html"
      type="text/widget"
    />
    Secondary banner a third.
  </div>
</DocumentFragment>
`;

exports[`Merchandising Banner Component renders type:banner banner 1`] = `
<DocumentFragment>
  .emotion-0 iframe,
.emotion-0 img {
  max-width: 100%;
  height: auto;
}

<div
    class="ss__banner ss__banner--banner emotion-0"
  >
    <script
      data-banner-html="Secondary banner."
      data-banner-type="html"
      type="text/widget"
    />
    Secondary banner.
    <script
      data-banner-html="Secondary banner."
      data-banner-type="html"
      type="text/widget"
    />
    Secondary banner again.
    <script
      data-banner-html="Secondary banner."
      data-banner-type="html"
      type="text/widget"
    />
    Secondary banner a third.
  </div>
</DocumentFragment>
`;

exports[`Merchandising Banner Component renders type:footer banner 1`] = `
<DocumentFragment>
  .emotion-0 iframe,
.emotion-0 img {
  max-width: 100%;
  height: auto;
}

<div
    class="ss__banner ss__banner--footer emotion-0"
  >
    <script
      data-banner-html="footer banner."
      data-banner-type="html"
      type="text/widget"
    />
    footer banner.
    <script
      data-banner-html="footer banner."
      data-banner-type="html"
      type="text/widget"
    />
    footer banner again.
  </div>
</DocumentFragment>
`;

exports[`Merchandising Banner Component renders type:header banner 1`] = `
<DocumentFragment>
  .emotion-0 iframe,
.emotion-0 img {
  max-width: 100%;
  height: auto;
}

<div
    class="ss__banner ss__banner--header emotion-0"
  >
    <script
      data-banner-html="Primary banner."
      data-banner-type="html"
      type="text/widget"
    />
    Primary banner.
    <script
      data-banner-html="Primary banner."
      data-banner-type="html"
      type="text/widget"
    />
    Primary banner again.
  </div>
</DocumentFragment>
`;

exports[`Merchandising Banner Component renders with classname 1`] = `
<DocumentFragment>
  .emotion-0 iframe,
.emotion-0 img {
  max-width: 100%;
  height: auto;
}

<div
    class="ss__banner ss__banner--banner classy emotion-0"
  >
    <script
      data-banner-html="Secondary banner."
      data-banner-type="html"
      type="text/widget"
    />
    Secondary banner.
    <script
      data-banner-html="Secondary banner."
      data-banner-type="html"
      type="text/widget"
    />
    Secondary banner again.
    <script
      data-banner-html="Secondary banner."
      data-banner-type="html"
      type="text/widget"
    />
    Secondary banner a third.
  </div>
</DocumentFragment>
`;

exports[`Merchandising Banner Component uses bocachica theme 1`] = `
<DocumentFragment>
  .emotion-0 iframe,
.emotion-0 img {
  max-width: 100%;
  height: auto;
}

<div
    class="ss__banner ss__banner--banner emotion-0"
  >
    <script
      data-banner-html="Secondary banner."
      data-banner-type="html"
      type="text/widget"
    />
    Secondary banner.
    <script
      data-banner-html="Secondary banner."
      data-banner-type="html"
      type="text/widget"
    />
    Secondary banner again.
    <script
      data-banner-html="Secondary banner."
      data-banner-type="html"
      type="text/widget"
    />
    Secondary banner a third.
  </div>
</DocumentFragment>
`;
